packages/walletconnect-solana/src/adapter.ts

```diff
@@ -42,60 +42,59 @@ const getConnectParams = (chainId: WalletConnectChainID, pairingTopic?: string):
         this._network = config.network;
     }
 
     async connect(): Promise<WalletConnectWalletInit> {
-        const client = await WalletConnectClient.init(this._options);
-
-        const pairings = client.pairing.getAll({ active: true });
-        // Prototypically, the user should be prompted to either:
-        // - Connect to a previously active pairing
-        // - Choose a new pairing
-        // There doesn't seem to be a WalletConnect-provided UI for this like there exists for the QRCode modal, though,
-        // and pushing this into user-land would be way too much
-        // If we decide to try and pair automatically, the UI will hang waiting for a pairing that might not complete
-        // const lastActivePairing = pairings.length ? pairings[pairings.length - 1].topic : undefined;
-        const lastActivePairing = undefined;
-
-        const { uri, approval } = await client.connect(getConnectParams(this._network, lastActivePairing));
-
-        if (uri) {
-            QRCodeModal.open(uri, () => {
-                throw new QRCodeModalError();
-            });
+        const client = this._client ?? (await WalletConnectClient.init(this._options));
+        const sessions = client.find(getConnectParams(this._network)).filter((s) => s.acknowledged);
+        if (sessions.length) {
+            // select last matching session
+            this._session = sessions[sessions.length - 1];
+            // We assign this variable only after we're sure we've received approval
+            this._client = client;
+        } else {
+            const { uri, approval } = await client.connect(getConnectParams(this._network));
+            if (uri) {
+                QRCodeModal.open(uri, () => {
+                    throw new QRCodeModalError();
+                });
+            }
+
+            this._session = await approval();
+            // We assign this variable only after we're sure we've received approval
+            this._client = client;
+
+            QRCodeModal.close();
         }
 
-        this._session = await approval();
-        // We assign this variable only after we're sure we've received approval
-        this._client = client;
-
-        QRCodeModal.close();
-
         return {
             publicKey: this.publicKey,
         };
     }
 
     async disconnect() {
         if (this._client && this._session) {
-            return await this._client.disconnect({
+            await this._client.disconnect({
                 topic: this._session.topic,
                 reason: getSdkError('USER_DISCONNECTED'),
             });
+            this._session = undefined;
         } else {
             throw new ClientNotInitializedError();
         }
     }
 
     get client(): WalletConnectClient {
         if (this._client) {
-            return this._client;
+            // TODO: using client.off throws an error
+            return Object.assign({}, this._client, { off: this._client.removeListener });
```

Comment:
hmm, `off` should be defined on the `@walletconnect/sign-client`: https://github.com/WalletConnect/walletconnect-monorepo/blob/6baeea8b925158c4e956c346857b8a96ddd9be74/packages/sign-client/src/client.ts#L72-L74

need to reproduce this one

Reply:
I saw this here: https://github.com/blockworks-foundation/mango-ui-v3/pull/397/files (happens even without this change)

Reply:
preview: https://mango-ui-v3-git-fork-amper-fb-wall-246ed5-blockworks-foundation.vercel.app/

Reply:
able to reproduce locally, too. this is an upstream issue; the distributed version of the `events` library that `@walletconnect/sign-client` bundles looks like the `.off` function has been tree-shaken out. will leave an issue to sort that out with them, but for now this is a suitable workaround